payload/vanguard/handler/upload.py

```diff
@@ -21,34 +23,84 @@ def parse_message(self, msg):
         prog_name_length = unpacked_data[3]
         prog_data_length = unpacked_data[4]
 
-        fmt = fmt + str(prog_name_length) + 's' + str(prog_data_length-1) + 's'
+        fmt = fmt + str(prog_name_length) + 's' + str(prog_data_length) + 's'
         second_unpacked_data  = struct.unpack_from(fmt, msg.message_data)
 
         self.program_name = second_unpacked_data[5]
         self.filename = self.program_name + '.js'
         self.program_data = second_unpacked_data[6]
+        self.directory = './uploads/programs/' + self.program_name + '/'
+        self.handle_chunk()
 
-        with open(self.filename,'a') as output_file:
-            output_file.write(self.program_data)  
+    def handle_chunk(self):
+        fmt_str = '!' + str(self.chunk_count) + '?'
 
+        if not os.path.exists(self.directory): # received first chunk of a new program 
+            os.makedirs(self.directory)
+            bool_arr = [False] * self.chunk_count
+            bool_arr[self.chunk - 1] = True
+            with open(self.directory + self.program_name + '.kbf', 'w+') as indexFile: #.kbf-Kubos Binary Format
+                indexFile.write(struct.pack(fmt_str, *bool_arr))
+            if self.chunk_count == 1:
+                self.store_chunk()
+                self.assemble_file()
+        else:  #This chunk is from a partially received program
+            index_struct = struct.Struct(fmt_str)
+            contents = []
+            with open(self.directory + self.program_name + '.kbf', 'rb') as indexfile:
+                while True:
+                    buf = indexfile.read(index_struct.size)
+                    if len(buf) != index_struct.size:
+                        break
+                    contents.append(index_struct.unpack_from(buf))
```

Comment:
I think you want to just concat these two lists together instead of appending, that way `contents` stays flat. in fact, it looks like you're plucking out `contents[0]` below to solve this issue :). in a nutshell, just use `extend` instead of `append` here, and you should be able to remove line 56